pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java

```diff
@@ -551,5 +585,16 @@ void refreshSegment(String segment) {
       }
       return _instanceSelector.select(brokerRequest, selectedSegments);
     }
+
+    boolean containsNoReplicaSegments(BrokerRequest brokerRequest) {
+      if (_noReplicaSegments.isEmpty()) {
+        return false;
+      }
+      List<String> selectedSegments = new ArrayList<>(_noReplicaSegments);
```

Comment:
I am not sure I follow this completely. 

1. Why do we have to invoke SegmentPruner to do this?

2. The SegmentPruner API takes the second arg as the segments "to be queried" by the broker request and returns a list of selected segments which can be a subset of passed list. Here we pass the "segmentsWithNoReplicas" list to pruner. IIUC, the API will return a subset of this list which is not what we want. We want to remove the "segmentsWithNoReplicas" from the list of segments to be queried by the broker request. I don't think that is happening here.

Reply:
My bad. I see what this function is returning. Ignore point 2

Reply:
A comment before the for loop will be really useful. I had a the same exact question as Sidd did until I went through it a few times.
Basically, we are trying to see if the unavailable segments would never have been hit by a query. If that is the case, then we are good for this query (e.g. in a partitioned use case, we may prune the unavailable segments).
Please add this as a comment block.

Reply:
Comment added.